DolbyIO/Source/SdkAccessModule/Private/DeviceManagement.cpp

```diff
@@ -3,132 +3,279 @@
 	}
 
-	void FDeviceManagement::GetAllDevices(FDvcDeviceManagement& DeviceManagement)
+	void FDeviceManagement::GetAllDevices()
 	{
 		DeviceManagement.get_audio_devices()
 		    .then(
-		        [this](auto&& Devices)
+		        [this](auto&& DvcDevices)
 		        {
-			        FDevices::FDvcDevices InitInputDevices;
-			        FDevices::FDvcDevices InitOutputDevices;
-			        FDevices::FDeviceNames InitInputDeviceNames;
-			        FDevices::FDeviceNames InitOutputDeviceNames;
-			        for (const auto& Device : Devices)
 			        {
-				        const auto DeviceName = ToFText(Device.name());
-				        if (IsInput(Device))
-				        {
-					        InitInputDevices.Add(Device);
-					        InitInputDeviceNames.Add(DeviceName);
-				        }
-				        if (IsOutput(Device))
-				        {
-					        InitOutputDevices.Add(Device);
-					        InitOutputDeviceNames.Add(DeviceName);
-				        }
+				        FScopeLock Lock{&AccessDevices};
+				        Devices.Append(DvcDevices.data(), DvcDevices.size());
+				        DLB_UE_LOG("In/Out-put %d devices added (%d in total)", DvcDevices.size(), Devices.Num());
 			        }
-			        InputDevices->Initialize(MoveTemp(InitInputDevices), MoveTemp(InitInputDeviceNames));
-			        OutputDevices->Initialize(MoveTemp(InitOutputDevices), MoveTemp(InitOutputDeviceNames));
+			        UpdateCurrentInputDeviceIndex();
+			        UpdateCurrentOutputDeviceIndex();
+			        Observer.OnListOfInputDevicesChanged();
+			        Observer.OnListOfOutputDevicesChanged();
 		        })
-		    .on_error(DLB_HANDLE_ASYNC_EXCEPTION);
+		    .on_error(MakeHandler(__LINE__));
+	}
+
+	FDeviceNames FDeviceManagement::GetDeviceNames(const EDirection direction) const
+	{
+		FDeviceNames Names;
+		FScopeLock Lock{&AccessDevices};
+		for (const auto& Device : Devices)
+		{
+			if (Device.direction() & direction)
+			{
+				Names.Add(ToFText(Device.name()));
+			}
+		}
+		return Names;
+	}
+
+	const FDvcDevice& FDeviceManagement::GetDeviceAt(const EDirection direction, Index Index) const
+	{
+		for (const auto& Device : Devices) // Devices are locked at caller
+		{
+			if (Device.direction() & direction)
+			{
+				if (Index-- == 0)
+				{
+					return Device;
+				}
+			}
+		}
+		throw std::logic_error{std::string{"Wrong "} + (IsInput(direction) ? "input" : "output") + " device Index"};
+	}
+
+	Index FDeviceManagement::GetNumberOfDevices(const EDirection direction) const
+	{
+		Index Count{0};
+		FScopeLock Lock{&AccessDevices};
+		for (const auto& Device : Devices)
+		{
+			if (Device.direction() & direction)
+			{
+				++Count;
+			}
+		}
+		return Count;
+	}
+
+	Index FDeviceManagement::GetDeviceIndex(const EDirection direction, const FDvcDevice& GivenDevice) const
+	{
+		Index Index = 0;
+		FScopeLock Lock{&AccessDevices};
+		for (const auto& Device : Devices)
+		{
+			if (Device == GivenDevice)
+			{
+				return Index;
+			}
+			if (Device.direction() & direction)
```

Comment:
This is also some sort of "find index by predicate" function that should be already in Unreal.